src/modinv64_impl.h

```diff
@@ -200,17 +200,128 @@
+ * all the time. Let's start by inlining div2n into update_de, and dropping the modulus operation at
+ * the end:
+ *
+ * def update_de(d, e, t, modulus, modulus_inv2n):
+ *     """Multiply matrix t/2^N with [d, e], modulo modulus."""
+ *     u, v, q, r = t
+ *     cd, ce = u * d + v * e, q * d + r * e
+ *     # Cancel out bottom N bits of cd and ce.
+ *     md = -((modulus_inv2n * (cd % (2**N))) % (2**N))
+ *     me = -((modulus_inv2n * (ce % (2**N))) % (2**N))
+ *     cd += md * modulus
+ *     ce += me * modulus
+ *     # And cleanly divide by 2**N.
+ *     return cd >> N, ce >> N
+ *
+ * Let's look at bounds on the ranges of these numbers. It can be shown that |u|+|v| and |q|+|r|
+ * never exceed 2^N, because a divstep's output f and g cannot exceed the input f and g in absolute
+ * value (see paragraph 8.3 in the paper). Assume that the input d and e are known to be in range
+ * (l*modulus,h*modulus) for some numbers l<0 and h>0. The bounds for |u|+|v| and |q|+|r| then imply
+ * that cd and ce (before cancelling out their bottom bits) are in range
+ * (-s*2^N*modulus,s*2^N*modulus), where s=max(|l|,|h|). After subtracting up to 2^N-1 times the
+ * modulus to cancel out bits, and dividing by 2^N at the end, we conclude that d and e have an
+ * output range slightly smaller than (-(s+1)*modulus,s*modulus).
+ *
+ * This means that every consecutive update_de increases the potential range of d and e. This can
+ * easily be counteracted by incrementing d and e by modulus whenever they're negative:
+ *
+ *     ...
+ *     if d < 0:
+ *         d += modulus
+ *     if e < 0:
+ *         e += modulus
+ *     cd, ce = u * d + v * e, q * d + r * e
+ *     # Cancel out bottom N bits of cd and ce.
+ *     ...
+ *
+ * Redoing the earlier bounds analysis for this code we get the same except now s=max(|l+1|,|h|),
+ * which means that l=-2 and h=1 results in s=1, resulting in outputs with the same bounds as the
+ * inputs.
+ *
+ * Note that increasing d by modulus is equal to incrementing cd by u*modulus and ce by q*modulus.
+ * The same is true for e and increments of v*modulus and r*modulus. So we could instead write:
+ *
+ *     ...
+ *     cd, ce = u * d + v * e, q * d + r * e
+ *     if d < 0:
+ *         cd, ce = cd + u*modulus, ce + q*modulus
+ *     if e < 0:
+ *         cd, ce = cd + v*modulus, ce + r*modulus
+ *     # Cancel out bottom N bits of cd and ce.
+ *     ...
+ *
+ * Now notice that we have two steps of corrections to cd and ce that add multiples of modulus.
+ * The second one depends on the first one, but they can still be efficiently combined by only
+ * computing the bottom bits of cd and ce at first, and using that to compute the final md, me
+ * values:
+ *
+ * def update_de(d, e, t, modulus, modulus_inv2n):
+ *     """Multiply matrix t/2^N with [d, e], modulo modulus."""
+ *     u, v, q, r = t
+ *     md, me = 0, 0
+ *     if d < 0:
+ *         md, me = md + u, me + q
+ *     if e < 0:
+ *         md, me = md + q, me + r
```

Comment:
I think this should be `md + v` instead of `md + q`?

Reply:
Nice catch, fixed.